gtsfm/retriever/netvlad_retriever.py

```diff
@@ -0,0 +1,175 @@
+
+Note: Similarity computation based off of Paul-Edouard Sarlin's HLOC:
+Reference: https://github.com/cvg/Hierarchical-Localization/blob/master/hloc/pairs_from_retrieval.py
+https://openaccess.thecvf.com/content_cvpr_2016/papers/Arandjelovic_NetVLAD_CNN_Architecture_CVPR_2016_paper.pdf
+
+Authors: John Lambert
+"""
+
+import math
+import os
+from pathlib import Path
+from typing import List, Optional, Tuple
+
+import matplotlib.pyplot as plt
+import numpy as np
+import torch
+
+import gtsfm.utils.logger as logger_utils
+from gtsfm.loader.loader_base import LoaderBase
+from gtsfm.retriever.retriever_base import RetrieverBase
+from gtsfm.frontend.cacher.global_descriptor_cacher import GlobalDescriptorCacher
+from gtsfm.frontend.global_descriptor.netvlad_global_descriptor import NetVLADGlobalDescriptor
+
+
+logger = logger_utils.get_logger()
+
+PLOT_SAVE_DIR = Path(__file__).parent.parent.parent / "plots"
+
+MAX_NUM_IMAGES = 10000
+
+
+class NetVLADRetriever(RetrieverBase):
+    def __init__(self, num_matched: int, blocksize: int = 10) -> None:
+        """
+        Args:
+            num_matched: number of K potential matches to provide per query. These are the top "K" matches per query.
+            blocksize: size of matching sub-blocks when creating similarity matrix.
+        """
+        self._num_matched = num_matched
+        self._global_descriptor_model = GlobalDescriptorCacher(global_descriptor_obj=NetVLADGlobalDescriptor())
+        self._blocksize = blocksize
+        self._min_score = 0.5
+
+    def run(self, loader: LoaderBase, visualize: bool = True) -> List[Tuple[int, int]]:
+        """Compute potential image pairs.
+
+        Args:
+            loader: image loader. The length of this loader will provide the total number of images
+                for exhaustive global descriptor matching.
+
+        Return:
+            pair_indices: (i1,i2) image pairs.
+        """
+        num_images = len(loader)
+
+        sim = self.compute_similarity_matrix(loader, num_images)
+
+        query_names = loader.image_filenames()
+        # Avoid self-matching and disallow lower triangular portion
+        is_invalid_mat = ~np.triu(np.ones((num_images, num_images), dtype=bool))
+        np.fill_diagonal(a=is_invalid_mat, val=True)
+        pairs = pairs_from_score_matrix(
+            sim, invalid=is_invalid_mat, num_select=self._num_matched, min_score=self._min_score
+        )
+
+        if visualize:
+            plt.imshow(np.triu(sim.detach().cpu().numpy()))
+            os.makedirs(PLOT_SAVE_DIR, exist_ok=True)
+            plt.title("Image Similarity Matrix")
+            np.savetxt(
+                fname=os.path.join(PLOT_SAVE_DIR, "netvlad_similarity_matrix.txt"),
+                X=sim.detach().cpu().numpy(),
+                fmt="%.2f",
+                delimiter=",",
+            )
+            plt.savefig(os.path.join(PLOT_SAVE_DIR, "netvlad_similarity_matrix.jpg"), dpi=500)
```

Comment:
Unrelated to this PR, but maybe we could generate a similar figure after the TwoViewEstimator where the z-axis is the pose error?

Reply:
that's a good idea, would be an interesting visualization